js/src/util/bn.ts

```diff
@@ -91,15 +91,25 @@ export function bigNumToNumber<T extends BN<BigNumArray>>(bn: T, scale?: number)
             number |= word * (BigInt(1) << BigInt(64 * i++));
         }
     }
-    if (typeof scale === 'number') {
-        const denominator = BigInt(Math.pow(10, scale));
+    if (typeof scale === 'number' && scale > 0) {
+        const denominator = BigInt('1' + '0'.repeat(scale));
         const quotient = number / denominator;
-        const remainder = number % denominator;
-        return bigIntToNumber(quotient) + (bigIntToNumber(remainder) / bigIntToNumber(denominator));
+        const remainder = negative? -(number % denominator) : number % denominator;
+        const integerPart = bigIntToNumber(quotient);
+        const fractionPart = padStart((remainder).toString(), scale);
```

Comment:
Minor suggestion, but we should be able to use the String built-in and remove the `padStart` below this:

```suggestion
        const fractionPart = remainder.toString().padStart(scale, '0');
```